core/drand_beacon.go

```diff
@@ -0,0 +1,352 @@
+package core
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/drand/drand/common"
+
+	"github.com/drand/drand/chain"
+	"github.com/drand/drand/chain/boltdb"
+	"github.com/drand/drand/fs"
+
+	"github.com/drand/drand/chain/beacon"
+	"github.com/drand/drand/key"
+	"github.com/drand/drand/log"
+	"github.com/drand/drand/net"
+	"github.com/drand/kyber/share/dkg"
+)
+
+// BeaconProcess is the main logic of the program. It reads the keys / group file, it
+// can start the DKG, read/write shars to files and can initiate/respond to TBlS
+// signature requests.
+type BeaconProcess struct {
+	opts *Config
+	priv *key.Pair
+	// current group this drand node is using
+	group *key.Group
+	index int
+
+	store       key.Store
+	privGateway *net.PrivateGateway
+	pubGateway  *net.PublicGateway
+	control     net.ControlListener
+
+	beacon *beacon.Handler
+
+	// dkg private share. can be nil if dkg not finished yet.
+	share   *key.Share
+	dkgDone bool
+
+	// manager is created and destroyed during a setup phase
+	manager  *setupManager
+	receiver *setupReceiver
+
+	// dkgInfo contains all the information related to an upcoming or in
+	// progress dkg protocol. It is nil for the rest of the time.
+	dkgInfo *dkgInfo
+	// general logger
+	log log.Logger
+
+	// global state lock
+	state  sync.Mutex
+	exitCh chan bool
+
+	// that cancel function is set when the drand process is following a chain
+	// but not participating. Drand calls the cancel func when the node
+	// participates to a resharing.
+	syncerCancel context.CancelFunc
+
+	// only used for testing currently
+	// XXX need boundaries between gRPC and control plane such that we can give
+	// a list of paramteres at each DKG (inluding this callback)
+	setupCB func(*key.Group)
+
+	// version indicates the base code variant
+	version common.Version
+
+	// only used for testing at the moment - may be useful later
+	// to pinpoint the exact messages from all nodes during dkg
+	dkgBoardSetup func(Broadcast) Broadcast
+}
+
+func NewBeaconProcess(log log.Logger, version common.Version, store key.Store,
```

Comment:
While the "functionality" is there, like I mentioned during our previous call, I find this a bit problematic because logic is very tied between "BeaconProcess" and "drand daemon". What seems to be a good way forward to decouple the two and give better abstractions is to move BeaconProcess to be an Go interface "Protocol" or stg:
* Creating a "Protocol" should only take what is the minimal amount of information needed generically, like the logger , the version, their folder and that's about it. 
* The instantiation should look into the folder and see what is inside and instantiate a key.Store if needed (or not !) 
* The Protocol should know itself whether it should already be running the beacon or not, drand daemon shouldn't be aware of the internals of the status of the Protocol. (it's just a proxy now).

I am "OK" to go forward with this PR as it is now, but I would appreciate some discussions around it or an issue about it. This coupling of logic makes me uneasy.

Reply:
ok, I've created an [issue](https://github.com/drand/drand/issues/860) for tracking this